### ISSUE TITLE
NaN metric propagation in GRPOTrainer with sparse multi-reward training

### ISSUE THREAD

**Issue by shriramc1** [huggingface/trl#4501]
## Bug Description

When using sparse multi-reward training with `logging_steps > 1`, metrics accumulate across steps. If a reward function is not active in some steps (returns `None`), it appends NaN values to the metrics list. The current implementation in `GRPOTrainer.log()` at line 1920 uses `sum(val) / len(val)` to compute the mean, which returns NaN if ANY value in the list is NaN, causing all valid metrics to be lost.

## Bug Location

https://github.com/huggingface/trl/blob/20760ba3ac7092432f67a5e3e9b2c624f168fdb0/trl/trainer/grpo_trainer.py#L1920

## Proposed Fix
```python
metrics = {
    key: torch.tensor(val).nanmean().item() if len(val) > 0 else 0.0
    for key, val in self._metrics[mode].items()
}
```

## Concrete Example

Consider a multi-task training setup with `logging_steps=4`:
- Step 1: Task A active, Task B inactive → metrics: `[0.5, NaN]`
- Step 2: Task A inactive, Task B active → metrics: `[0.5, NaN, NaN, 0.8]`
- Step 3: Both active → metrics: `[0.5, NaN, NaN, 0.8, 0.6, 0.7]`
- Step 4 (logging): `sum([0.5, NaN, NaN, 0.8, 0.6, 0.7]) / 6 = NaN` ❌

All valid reward values are lost due to NaN propagation.

## Minimal Reproducible Example

```python
from trl import GRPOTrainer, GRPOConfig
from datasets import Dataset

# Create dataset with mixed tasks
dataset = Dataset.from_dict({
    "prompt": ["task_a question", "task_b question", "task_a question", "both tasks"]
})

# Multi-reward functions that return None for non-applicable samples
def reward_func_a(prompts, completions, **kwargs):
    return [1.0 if "task_a" in p else None for p in prompts]

def reward_func_b(prompts, completions, **kwargs):
    return [0.8 if "task_b" in p else None for p in prompts]

trainer = GRPOTrainer(
    model="Qwen/Qwen2-0.5B-Instruct",
    reward_funcs=[reward_func_a, reward_func_b],
    args=GRPOConfig(
        output_dir="./output",
        logging_steps=4,  # Accumulate metrics across 4 steps
    ),
    train_dataset=dataset,
)

trainer.train()
# After 4 steps, logged metrics will be NaN even though valid values exist

### System Info

- Platform: Linux-6.12.31-35.92.amzn2023.x86_64-x86_64-with-glibc2.34
- Python version: 3.12.9
- TRL version: 0.24.0
- PyTorch version: 2.8.0
- accelerator(s): cpu
- Transformers version: 4.57.1
- Accelerate version: 1.10.1
- Accelerate config: not found
- Datasets version: 4.2.0
- HF Hub version: 0.35.3
- bitsandbytes version: not installed
- DeepSpeed version: 0.18.0
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 2.4.0
- PEFT version: 0.17.1
- vLLM version: 0.10.2

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by pramodith** [comment]
Is there a benefit of returning `None` instead of `0` when a task is unrelated?

**Comment by qgallouedec** [comment]
None are counted as 0: 
https://github.com/huggingface/trl/blob/04f2b4db2dc6c52ee288067bf36d7a80e17ca213/trl/trainer/grpo_trainer.py#L1606

so technically you could just use 0. But returning None is actually a better practice because it allows to catch and issue a warning if needed:

https://github.com/huggingface/trl/blob/04f2b4db2dc6c52ee288067bf36d7a80e17ca213/trl/trainer/grpo_trainer.py#L1126-L1137

**Comment by shriramc1** [comment]
> Is there a benefit of returning `None` instead of `0` when a task is unrelated?

Well returning 0 would mess up the advantage calculation. We avoid including a reward by setting it to `None`

**Comment by pramodith** [comment]
> > Is there a benefit of returning `None` instead of `0` when a task is unrelated?
> 
> Well returning 0 would mess up the advantage calculation. We avoid including a reward by setting it to `None`

@shriramc1 as Quentin points out in the code snippet above the reward aggregation step treats `None` as 0s anyway. In general the rewards are just summed up, and the std is computed over the final summed rewards so I don't think the rewards would be messed up in any way by returning `0` instead of `None`.

**Comment by behroozazarkhalili** [comment]
This looks already fixed on current `main`. `GRPOTrainer.log()` now filters out NaN values before averaging, which addresses exactly the contamination you described:

```python
for key, val in self._metrics[mode].items():
    # Filter out NaN values before averaging. A reward function that returns None for all samples
    # in a batch produces NaN for that batch's metric. With logging_steps > 1, a naive sum()/len()
    # would let a single NaN contaminate valid data from other batches. Only return None when no
    # valid values remain (e.g. JSON loggers crash on float NaN).
    valid = [v for v in val if not math.isnan(v)]
    metrics[key] = sum(valid) / len(valid) if valid else None
```

Running your concrete example through this:

```
val = [0.5, NaN, NaN, 0.8, 0.6, 0.7]
-> valid = [0.5, 0.8, 0.6, 0.7]  ->  mean = 0.65   (the old sum()/len() gave NaN)
```

Note it also handles the all-NaN case by returning `None` rather than `NaN`, which avoids the JSON-logger crash that a plain `nanmean()` would still produce.

This landed in #5235. I think this can be closed as resolved — happy to be corrected if you still see NaN propagation on a recent version.
